### PR TITLE
update GnuPG and friends to 2.4.0

### PR DESCRIPTION
**Pull request by saper**
gnupg 2.2.21 -> 2.4.0
libgcrypt 1.8.6 -> 1.10.1
libassuan 2.5.3 -> 2.5.5
libksba 1.4.0 -> 1.6.3
libgpg-error 1.46

### CODE REVIEW

**Comment by tlaurion** [comment]
GPGRT_CONFIG is not found in libgcrypt configure script...

**Comment by tlaurion** [comment]
@saper This is the issue where solution is not working 
https://dev.gnupg.org/T6257

**Comment by tlaurion** [comment]
We rely on pkg-config to provide proper paths, where as of now, libgpg-error.so is found, but libgpg-error.so.1 is searched in wrong path and fails linking, both locally and under CircleCI (debian-11 based for build reproducibility (not to be confounded with reproducibility of builds. We just want a setup to recommend that works for all, in the meantime of having a build stack that produces same binaries).

I was trying to fix this and poing you again to gpg-card ninary addition (new from 2.3+) which needs to be explored to hopefully remove the screenscaping, that as you seen, is fragile and does not survive version bumps.

**Comment by tlaurion** [comment]
Updated https://github.com/osresearch/heads/pull/1350#issuecomment-1483140260

**Comment by tlaurion** [comment]
> @saper fails at https://app.circleci.com/pipelines/github/tlaurion/heads/1561/workflows/933c2b4e-c532-4166-bec5-2be8ca9166f2/jobs/19454?invite=true#step-103-1904

@saper if you had a previous version of libgpgerror where searched for, that is why it works in your build env but that lucky build cannot be made again. I wiped install/x86, had to rebuild complaining modules (eg: make BOARD=x230-hotp-maximized libgpg-error.clean) to reproduce same thing that CircleCI complains about.

**Comment by tlaurion** [comment]
https://app.circleci.com/pipelines/github/tlaurion/heads?branch=gnupg-2.4.0-update
Will try to talk to gniibe

**Comment by tlaurion** [comment]
Seems like rhe issue are with tests not reusing same toolchain and libraries used for compilation.

**Comment by saper** [comment]
I have disconnected the tests from the libgcrypt build with https://github.com/osresearch/heads/pull/1350/commits/924f78fe9301bd91b6abe10746d7c3215c7c8043

**Comment by saper** [comment]
Force pushed the last commit with a whitespace fix by @tlaurion

**Comment by tlaurion** [comment]
Note that ppc64 related patches are gone, will test those in next batch of tests, hopefully I will remember.

**Comment by tlaurion** [comment]
Some notes on size changes (produced by playing with initrd.cpio.xz under /tmp dir), comparing output of
`xz -d initrd.cpio.xz ; cpio -i < initrd.cpio ; find . -type f -ls | sort -r -n -k7 |grep -v cpio | awk -F " " {'print $NF " " $7'} | sort > output `

```
--- master-hotp-fbwhiptail/output	2023-03-27 14:02:18.647000000 -0400
+++ gpg24-hotp-fbwhiptail/output	2023-03-27 14:05:20.344000000 -0400
@@ -18,9 +18,12 @@
 ./bin/flash.sh 6653
 ./bin/flashtool 18336
 ./bin/generic-init 1345
-./bin/gpg 751456
-./bin/gpg-agent 314584
+./bin/gpg 828136
+./bin/gpg-agent 343984
+./bin/gpg-card 205704
+./bin/gpg-connect-agent 119216
 ./bin/gpg-gui.sh 9056
+./bin/gpgsm 442792
 ./bin/gpgv 150
 ./bin/gui-init 21584
 ./bin/hotp 5912
@@ -53,7 +56,7 @@
 ./bin/qrenc 5904
 ./bin/qubes-measure-luks 597
 ./bin/reboot 375
-./bin/scdaemon 346376
+./bin/scdaemon 450792
 ./bin/scp 22664
 ./bin/seal-hotpkey 3328
 ./bin/seal-totp 1545
@@ -72,7 +75,7 @@
 ./bin/whiptail 35
 ./bin/xx30-flash.init 639
 ./etc/ash_functions 1930
-./etc/config 625
+./etc/config 771
 ./etc/distro/keys/archlinux.key 1725
 ./etc/distro/keys/qubes-4.key 1629
 ./etc/distro/keys/qubes-testing.key 3078
@@ -95,11 +98,11 @@
 ./lib/libcryptsetup.so.12 403152
 ./lib/libc.so 596544
 ./lib/libdevmapper.so.1.02 284184
-./lib/libgcrypt.so.20 1054216
-./lib/libgpg-error.so.0 109520
+./lib/libgcrypt.so.20 776328
+./lib/libgpg-error.so.0 121704
 ./lib/libjson-c.so.5 81024
-./lib/libksba.so.8 199016
-./lib/libmbedcrypto.so.0 288272
+./lib/libksba.so.8 207216
+./lib/libmbedcrypto.so.0 325104
 ./lib/libnpth.so.0 14552
 ./lib/libpci.so.3 52272
 ./lib/libpci.so.3.5.4 52272
@@ -109,15 +112,18 @@
 ./lib/libqrencode.so.3 39232
 ./lib/libtpm.so 314728
 ./lib/libusb-0.1.so.4 18440
-./lib/libusb-1.0.so.0 84584
+./lib/libusb-1.0.so.0 92808
 ./lib/libuuid.so.1 14720
 ./lib/libz.so.1 108832
-./lib/modules/e1000e.ko 304272
-./lib/modules/ehci-hcd.ko 64800
-./lib/modules/ehci-pci.ko 10728
-./lib/modules/usb-storage.ko 132360
-./lib/modules/xhci-hcd.ko 159840
-./lib/modules/xhci-pci.ko 10568
+./lib/modules/e1000.ko 223136
+./lib/modules/ehci-hcd.ko 116912
+./lib/modules/ehci-pci.ko 12312
+./lib/modules/ohci-hcd.ko 73280
+./lib/modules/ohci-pci.ko 13896
+./lib/modules/uhci-hcd.ko 61320
+./lib/modules/usb-storage.ko 140888
+./lib/modules/xhci-hcd.ko 276096
+./lib/modules/xhci-pci.ko 13448
 ./mount-boot 1375
 ./output 0
 ./run/cryptsetup/.placeholder 0
```

**Comment by tlaurion** [comment]
Also note that neither gpg-connect-agent, gpgsm nor gpg-card are needed, but sizes above show respective added space consomption outside of initrd.cpio.xz being compressed (which can be obtained by comparing coreboot stitching output in logs of CircleCI output, and where addition of those 3 breaks builds of legacy boards and t520 board as of now).

I would advise removing gpg-connect-agent as well unless really necessary.
Screen scraping haven't changed that much, where next step on my side will be change functions under oem-factory-reset and separate them to more properly call key-attr changes, define type of key desired etc, where first step and goal on my side is prototype this to be useable to recreate current RSA 3072 keysize without breaking anything and do the same in ram and then key to card, and move from there.

**Comment by tlaurion** [comment]
Suppressed things further for gpg under https://github.com/tlaurion/heads/tree/gnupg-2.4.0-update

**Comment by tlaurion** [comment]
> Work in progress, smartcards do not seem to work

@saper what do you mean?

**Comment by tlaurion** [comment]
@saper 

Also note that the following would be needed for the factory reset/reownership wizard to work with 2.4 

https://github.com/tlaurion/heads/commit/207c8b5a8a1ea39a78f688ab0a953ed11ac7478d

**Comment by tlaurion** [comment]
> Note that ppc64 related patches are gone, will test those in next batch of tests, hopefully I will remember.

@saper damnit I forgot.
Any reason why that patch was removed?

File: patches/libgpg-error-1.37.patch
Comment (tlaurion) [inline patches/libgpg-error-1.37.patch:1]: Basically: why this patch has been removed  for ppc64 @saper?

**Review by tlaurion — CHANGES_REQUESTED** [review]
If you have a quick answer on that it would be nice, otherwise will need to test.

File: patches/libgpg-error-1.37.patch
Comment (saper) [reply patches/libgpg-error-1.37.patch:1]: No reason, I probably made a mistake and though this has already been included upstream.

**Comment by tlaurion** [comment]
To be tested on top of #1381

**Comment by saper** [comment]
I wouldn't hold my breath :(

File: modules/gpg2 [hunk 1]
```diff
@@ -53,6 +54,6 @@ gpg2_target := $(MAKE_JOBS) \
 		DESTDIR="$(INSTALL)" \
 		install
 
-gpg2_output := g10/gpg agent/gpg-agent scd/scdaemon
+gpg2_output := g10/gpg agent/gpg-agent scd/scdaemon tools/gpg-connect-agent
```
Comment (tlaurion) [inline modules/gpg2:57]: Se could remove tools/gpg-connect-agent

Comment (tlaurion) [inline modules/gpg2:57]: We should revert adding tools/gpg-connect-agent

File: modules/libassuan [hunk 2]
```diff
@@ -13,7 +13,6 @@ libassuan_configure := \
 	--host $(MUSL_ARCH)-linux-musl \
 	--prefix "/" \
 	--disable-static \
-	--disable-debug \
```
Comment (tlaurion) [inline modules/libassuan:16]: We should revert removing --disable-debug

File: modules/gpg2 [hunk 3]
```diff
@@ -41,6 +41,7 @@ gpg2_configure := \
 	--disable-wks-tools \
 	--disable-gnutls \
 	--disable-dirmngr \
+	--disable-ntbtls \
```
Comment (tlaurion) [inline modules/gpg2:44]: Could add 
--disable-libdns

File: modules/gpg2 [hunk 4]
```diff
@@ -1,10 +1,10 @@
 modules-$(CONFIG_GPG2) += gpg2
 
-gpg2_version := 2.2.21
+gpg2_version := 2.4.0
```
Comment (tlaurion) [inline modules/gpg2:3]: 2.4.1 fixes issues with copy to card for yubikey.
2.4.2 is out.

**Comment by tlaurion** [comment]
Merged under #1422 with minor modification

**Pull request closed without merging**